apps/src/lib/kits/maker/ui/BoardSetupCheck.jsx

```diff
@@ -0,0 +1,335 @@
+/** @file Maker Board setup checker */
+import React, {Component} from 'react';
+import CircuitPlaygroundBoard from '../CircuitPlaygroundBoard';
+import {ensureAppInstalled, findPortWithViableDevice} from '../portScanning';
+import SetupStep, {
+  HIDDEN,
+  WAITING,
+  ATTEMPTING,
+  SUCCEEDED,
+  FAILED,
+  CELEBRATING,
+} from './SetupStep';
+
+const STATUS_IS_CHROME = 'statusIsChrome';
+const STATUS_APP_INSTALLED = 'statusAppInstalled';
+const STATUS_WINDOWS_DRIVERS = 'statusWindowsDrivers';
+const STATUS_BOARD_PLUG = 'statusBoardPlug';
+const STATUS_BOARD_CONNECT = 'statusBoardConnect';
+const STATUS_BOARD_COMPONENTS = 'statusBoardComponents';
+
+const initialState = {
+  isDetecting: false,
+  caughtError: null,
+  [STATUS_IS_CHROME]: WAITING,
+  [STATUS_APP_INSTALLED]: WAITING,
+  [STATUS_WINDOWS_DRIVERS]: WAITING,
+  [STATUS_BOARD_PLUG]: WAITING,
+  [STATUS_BOARD_CONNECT]: WAITING,
+  [STATUS_BOARD_COMPONENTS]: WAITING,
+};
+
+export default class BoardSetupCheck extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {...initialState};
+  }
+
+  hide(selector) {
+    this.setState({[selector]: HIDDEN});
+  }
+
+  fail(selector) {
+    this.setState({[selector]: FAILED});
+  }
+
+  spin(selector) {
+    this.setState({[selector]: ATTEMPTING});
+  }
+
+  succeed(selector) {
+    this.setState({[selector]: SUCCEEDED});
+  }
+
+  thumb(selector) {
+    this.setState({[selector]: CELEBRATING});
+  }
+
+  getSurveyURL() {
+    const baseFormURL = 'https://docs.google.com/forms/d/e/1FAIpQLSe4NB7weq20sydf4kKn3QzIIn1O91hfPNU0U6b2xc1W6w44eQ/viewform';
+    const userAgentFieldFill = `entry.1520933088=${encodeURIComponent(navigator.userAgent)}`;
+    const prettifiedCurrentStates = JSON.stringify(this.state, null, 2);
+    const setupStatesFieldFill = `entry.804069894=${encodeURIComponent(prettifiedCurrentStates)}`;
+    return `${baseFormURL}?${userAgentFieldFill}&${setupStatesFieldFill}`;
+  }
+
+  detect() {
```

Comment:
At least with some of them I think I've extracted the detection logic as far as I can, short of completely inverting control and making this a dumb component with some sort of parent driving the dozens of state changes it goes through while detecting.  For example:

```jsx
 detectChromeAppInstalled() {
    this.spin(STATUS_APP_INSTALLED);
    return promiseWaitFor(200)
        .then(ensureAppInstalled)
        .then(() => this.succeed(STATUS_APP_INSTALLED))
        .catch(error => {
          this.fail(STATUS_APP_INSTALLED);
          return Promise.reject(error);
        });
  }
```
This is pretty much all cosmetic / animation except the `.then(ensureAppInstalled)` call and the fact that we make sure to re-reject caught errors.

I can push further in this direction, possibly deduplicating these into a `detectionStep(stepKey, stepWork)` helper - or if you think it'd be easier, I can investigate a truly inverted setup with all connection/detection logic in its own module that frequently state changes to a pure display component as a side effect.  But I think the latter quite a bit of work at this point - it sounds like just switching to redux, which I'd so far assumed wasn't worth it for the setup page.